ui/store/action-queue/index.js

```diff
@@ -0,0 +1,201 @@
+  actionRetryQueue.length = 0;
+}
+
+// add action to queue
+const executeActionOrAddtoRetryQueue = (item) => {
+  if (actionRetryQueue.some((act) => act.actionId === item.actionId)) {
+    return;
+  }
+
+  if (background.connectionStream.readable) {
+    executeAction({
+      action: item,
+      disconnectSideeffect: () => actionRetryQueue.push(item),
+    });
+  } else {
+    actionRetryQueue.push(item);
+  }
+};
+
+/**
+ * Promise-style call to background method
+ * In MV2: invokes promisifiedBackground method directly.
+ * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
+ *  the queue is then immediately processed if background connection is available.
+ *  On completion (successful or error) the action is removed from the retry queue.
+ *
+ * @param {string} method - name of the background method
+ * @param {Array} [args] - arguments to that method, if any
+ * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue waiting for a retry.
+ * @returns {Promise}
+ */
+export function submitRequestToBackground(
+  method,
+  args = [],
+  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
+) {
+  if (isManifestV3) {
+    return new Promise((resolve, reject) => {
+      executeActionOrAddtoRetryQueue({
+        actionId,
+        request: { method, args },
+        resolve,
+        reject,
+      });
+    });
+  }
+  return promisifiedBackground[method](...args);
+}
+
+/**
+ * Callback-style call to background method
+ * In MV2: invokes promisifiedBackground method directly.
+ * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
+ *  the queue is then immediately processed if background connection is available.
+ *  On completion (successful or error) the action is removed from the retry queue.
+ *
+ * @param {string} method - name of the background method
+ * @param {Array} [args] - arguments to that method, if any
+ * @param callback - Node style (error, result) callback for finishing the operation
+ * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue.
+ */
+export const callBackgroundMethod = (
+  method,
+  args = [],
+  callback,
+  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
+) => {
+  if (isManifestV3) {
+    const resolve = (value) => callback(null, value);
+    const reject = (err) => callback(err);
+    executeActionOrAddtoRetryQueue({
+      actionId,
+      request: { method, args },
+      resolve,
+      reject,
+    });
+  } else {
+    background[method](...args, callback);
+  }
+};
+
+async function executeAction({ action, disconnectSideeffect }) {
+  const {
+    actionId,
+    request: { method, args },
+    resolve,
+    reject,
+  } = action;
+  try {
+    resolve(await promisifiedBackground[method](...args, actionId));
+  } catch (err) {
```

Comment:
@naugtur , I have added actionId here as the last argument, it is useful to make action call idempotent in some scenario.

Reply:
It crosses abstraction boundaries. Now all `background` functions have to understand they're being used that way. 
- How far does it reach? Is it used in the RPC client or in the background worker?
- Why is it only added in one place and not the others? Seems like an oversight.

IMHO, if we want to add it, it'd need to be the first argument and be made mandatory, otherwise it'll hurt us later.

Reply:
I used it here https://github.com/MetaMask/metamask-extension/pull/15667/files#diff-675c437262d906486cf5436c0b64535347286cc4d1081dbd863fb45b8a01b81aR755

I kept it last argument so that methods deal with it only if they need to. If we do not want to pass it to all background methods, I can pass it specifically where needed.

Reply:
Intuition tells me it's a footgun.

Current change visible here sends it or doesn't send it depending on the current state - you only added it in the queue processing but it's not defined if connection was available. 
So it only guarantees idempotence across multiple retries, but not between the initial call and first retry

Reply:
I changed it to pass actionId only in scenario needed.